Mathlib/GroupTheory/Coset.lean

```diff
@@ -0,0 +1,899 @@
+  simp [card_eq_card_quotient_mul_card_subgroup s, @dvd_mul_right ℕ]
+#align subgroup.card_quotient_dvd_card Subgroup.card_quotient_dvd_card
+#align add_subgroup.card_quotient_dvd_card AddSubgroup.card_quotient_dvd_card
+
+open Fintype
+
+variable {H : Type _} [Group H]
+
+@[to_additive]
+theorem card_dvd_of_injective [Fintype α] [Fintype H] (f : α →* H) (hf : Function.Injective f) :
+    card α ∣ card H := by
+  classical calc
+      card α = card (f.range : Subgroup H) := card_congr (Equiv.ofInjective f hf)
+      _ ∣ card H := card_subgroup_dvd_card _
+
+#align subgroup.card_dvd_of_injective Subgroup.card_dvd_of_injective
+#align add_subgroup.card_dvd_of_injective AddSubgroup.card_dvd_of_injective
+
+@[to_additive]
+theorem card_dvd_of_le {H K : Subgroup α} [Fintype H] [Fintype K] (hHK : H ≤ K) : card H ∣ card K :=
+  card_dvd_of_injective (inclusion hHK) (inclusion_injective hHK)
+#align subgroup.card_dvd_of_le Subgroup.card_dvd_of_le
+#align add_subgroup.card_dvd_of_le AddSubgroup.card_dvd_of_le
+
+@[to_additive]
+theorem card_comap_dvd_of_injective (K : Subgroup H) [Fintype K] (f : α →* H) [Fintype (K.comap f)]
+    (hf : Function.Injective f) : Fintype.card (K.comap f) ∣ Fintype.card K := by
+  haveI : Fintype ((K.comap f).map f) :=
+      Fintype.ofEquiv _ (equivMapOfInjective _ _ hf).toEquiv;
+    calc
+      Fintype.card (K.comap f) = Fintype.card ((K.comap f).map f) :=
+        Fintype.card_congr (equivMapOfInjective _ _ hf).toEquiv
+      _ ∣ Fintype.card K := card_dvd_of_le (map_comap_le _ _)
+
+#align subgroup.card_comap_dvd_of_injective Subgroup.card_comap_dvd_of_injective
+#align add_subgroup.card_comap_dvd_of_injective AddSubgroup.card_comap_dvd_of_injective
+
+end Subgroup
+
+namespace QuotientGroup
+
+variable [Group α]
+
+/-- If `s` is a subgroup of the group `α`, and `t` is a subset of `α ⧸ s`, then there is a
+(typically non-canonical) bijection between the preimage of `t` in `α` and the product `s × t`. -/
+@[to_additive "If `s` is a subgroup of the additive group `α`, and `t` is a subset of `α ⧸ s`, then
+ there is a (typically non-canonical) bijection between the preimage of `t` in `α` and the product
+ `s × t`."]
+noncomputable def preimageMkEquivSubgroupTimesSet (s : Subgroup α) (t : Set (α ⧸ s)) :
+    QuotientGroup.mk ⁻¹' t ≃ s × t
+    where
+  toFun a :=
+    ⟨⟨((Quotient.out' (QuotientGroup.mk a)) : α)⁻¹ * a,
+        leftRel_apply.mp (@Quotient.exact' _ (leftRel s) _ _ <| Quotient.out_eq' _)⟩,
+      ⟨QuotientGroup.mk a, a.2⟩⟩
+  invFun a :=
+    ⟨Quotient.out' a.2.1 * a.1.1,
+      show QuotientGroup.mk _ ∈ t by
+        rw [mk_mul_of_mem _ a.1.2, out_eq']
+        exact a.2.2⟩
+  left_inv := fun ⟨a, ha⟩ => Subtype.eq <| show _ * _ = a by simp
+  right_inv := fun ⟨⟨a, ha⟩, ⟨x, hx⟩⟩ => by ext <;> simp [ha]
+#align quotient_group.preimage_mk_equiv_subgroup_times_set
+  QuotientGroup.preimageMkEquivSubgroupTimesSet
+#align quotient_add_group.preimage_mk_equiv_add_subgroup_times_set
```

Comment:
I guess it's too late to rename this to `quotient_add_group.preimage_mk_equiv_add_subgroup_prod_set`

Reply:
Why would it be too late?

Reply:
Too late to do so without having to forward port is I guess what I meant.

Reply:
We can just rename that mathlib4 version and `#align` names.

Reply:
Ooh, you want to change the ml3 name? I don't see *any* value in that. But I do think we could rename the ml4 name right now.